### PR TITLE
Feature/oauth2 proxy

### PR DESCRIPTION
**Pull request by RichardLinde**
Added a general documentation for using a oauth2 proxy to secure services

### CODE REVIEW

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 1]
```diff
@@ -0,0 +1,70 @@
+:imagesdir: ../images
+= OAuth2 Proxy
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:2]: The level one heading should be more descriptive like "Authentication with Keycloak using a OAuth2 Proxy"

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+== Setup
+
+The OAuth proxy can be deployed from a link:https://github.com/OAuth2-proxy/OAuth2-proxy/releases/latest[prebuilt binary], a prebuilt link:https://quay.io/repository/OAuth2-proxy/OAuth2-proxy?tab=tags&tag=latest[docker image] or by link:https://github.com/OAuth2-proxy/manifests[helm chart] inside a Kubernetes cluster. It's recommended to use Kubernetes.
+
+image::OAuth2_Proxy-RequestFlow.drawio.svg[Request Flow]
+
+
+=== Keycloak 
+
+In this example, link:https://www.keycloak.org/[keycloak] is used as identity provider. For more details on keycloak see the devon link:https://devonfw.com/docs/java/current/cross_cutting/keycloak.html[documentation].
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:21]: Is there a keycloak documentation in devonfw? I'm not aware of it...

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+image::OAuth2_Proxy-RequestFlow.drawio.svg[Request Flow]
+
+
+=== Keycloak 
+
+In this example, link:https://www.keycloak.org/[keycloak] is used as identity provider. For more details on keycloak see the devon link:https://devonfw.com/docs/java/current/cross_cutting/keycloak.html[documentation]. 
+To deploy keycloak in the cluster, it's recommended to use the helm chart from link:https://artifacthub.io/packages/helm/bitnami/keycloak[bitnami]. To configure the keycloak deployment, use a link:https://github.com/bitnami/charts/blob/main/bitnami/keycloak/values.yaml[values.yaml] file. After keycloak is running go to the keycloak admin ui and configure a client. Make sure to follow the link:https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider#keycloak-oidc-auth-provider[guide of OAuth2] proxy to set up the group mapping properly.
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:22]: Sadly I'm not a keycloak expert. But I'm having a strange feeling just recommending the bitnami helm charts without any validation process of those helm charts.

Maybe just remove it and also say that keycloak offers various deployment options. I would not recommend anything here.
The rest is fine

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+----
+helm repo add bitnami https://charts.bitnami.com/bitnami
+
+helm upgrade --install keycloak bitnami/keycloak -f config/keyCloakValues.yaml
+----
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:28]: See above

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+=== Keycloak ingress
+
+This ingress is created by the keycloak helm chart, allowing a host like `keycloak.localtest.me` to be mapped to the keycloak service. This is important for setting up redirect URLs in the OAuth2 proxy.
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:32]: What is localtest.me. Why do I need it ;-) 

I see two options here:
1. Make clear that it's all about a local setup and explain the localtest.me aso.
2. Make it more abstract. "A ingress gateway is necessary to access keycload" -> Leading me to the question "Is really a ingress necessary? Or would a load balancer service be enough? If an ingress is necessary why?
It's important to make clear that we're now

Comment (baumeister25) [reply modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:32]: I think option two would be best and most helpful. Less details more conceptual description

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+=== Keycloak Database
+
+The keycloak database is created by the helm chart. If no database is connected to keycloak, a in memory database is used.
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:36]: For bitnami that's true.
Here I would just list the options. A in memory db cannot be the solution for a production cluster I guess.

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+=== Main ingress
+
+The main ingress only task is to redirect all requests from a host path like `localtest.me` to the OAuth2 proxy. Make sure to have a single ingress controller running in the cluster.
+
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:41]: Do you have multiple ingress? Or is there one ingress with different routes. That would be important. I would have expected only one ingress.

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+=== Auth2 Proxy
+
+The OAuth 2 proxy is running between the user and the application. If the user sends a request, the proxy will check the request for the proxy's session cookie or a JWT. When the authentication is not provided, the user's browser is redirected to the OAuth screen of a configured identity provider. If the request is authorized, the proxy will forward the request to a configured upstream.
+To set up the OAuth 2 proxy, create a deployment with the docker image. Inside the deployment file, link:https://OAuth2-proxy.github.io/OAuth2-proxy/docs/configuration/overview[configure] the OAuth proxy.
+
+=== Spring Server
+
+In this example, the secured application is a basic Spring Boot server with a simple API route to GET data.
+Because of the OAuth2 proxy, the server doesn't need to use Spring security to protect the application, keeping it simple and manageable.
```
Comment (baumeister25) [inline modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:50]: Could be any application. I would just name it Application

**Review by baumeister25 — CHANGES_REQUESTED** [review]
IMHO the solution should be on a more abstract level. 
In some parts it's very concrete (bitnami helm charts, localtest.me)

I think we should focus on 2 different things.
First: The overall concept. You sarted very good with the flow diagram. But you should also explain the diagram a bit and the flow without technical descriptions like deployments.

Then the different tools like you did: Keycloak, Oauth Proxies, Ingress
For each tool the configuration (as long as applicable for PROD and local) should be explained.
E.g. ingress routes (not the URLs, but what subroutes do i need and to which service do they direct) or e.g. the Oauth2 proxy settings. Where are they stored? How do they look like? Are they organized in a specific way or splitted, etc?

And last but not least as you already started a more concrete guide for the local deployment could be done. What is localTest.me, why do we need that. What about the specific configs, where to put them, etc.

File: modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc (hunk 5 above)
Comment (RichardLinde) [reply modules/ROOT/pages/cross_cutting/oauth2_proxy.adoc:32]: The keycloak ingress can be  replaced with an extra entry in the main ingress. I will remove the keycloak ingress and update the documentation.

**Comment by hohwille** [comment]
@RichardLinde thanks for this PR and your great contribution 👍 
@baumeister25 what is the status here?
IMHO @RichardLinde has addressed your concerns (However, there is still a slight reference to keycloak at the ingress section that is still not so clear to me).
What also confuses me is that there is a broken image link `image::OAuth2_Proxy-LocalProblem.drawio.svg[Localhost Problem]` but that is already present before and not related to this PR at all.
It would be great if we can get this PR merged as IMHO it is already on the finishing line.

**Review by baumeister25 — APPROVED** [review]
Thanks for your work. I really like this article as it includes the most important points and links!

**Pull request merged**